src/heaps/binomial_heap.jl

```diff
@@ -0,0 +1,368 @@
+    if length(_heap) == 0
+        return nothing
+    end
+    it = iterate(_heap)
+    @inbounds temp = _heap[1]
+    while it !== nothing
+        elem, state = it
+        if elem.data < temp.data 
+            temp = elem 
+        end
+        it = iterate(_heap, state) 
+    end
+    return (temp.data, temp)
+end
+
+function update_handle(c::Int, h::MutableBinomialHeap)
+    node::Union{Nothing,MutableBinomialHeapNode{T}} where T = nothing 
+    for i = 1:h.cumm_nodecount
+        node = nothing
+        try
+            node = findNode(h, i)
+            node.handle = node.handle + c
+        catch _
+            continue
+        end
+    end
+end
+
+function union!(h1::MutableBinomialHeap{T}, h2::MutableBinomialHeap{T}, flag::Bool) where T
+    _new = MutableBinomialHeap{T}()
+    if flag
+        update_handle(h1.cumm_nodecount, h2)
+    end
+    l1 = h1.rootList
+    l2 = h2.rootList
+    h1.nodecount += h2.nodecount
+    h1.cumm_nodecount += h2.cumm_nodecount
+    iter1 = iterate(l1)
+    iter2 = iterate(l2)
+    while (iter1 !== nothing && iter2 !== nothing)
+        (elem1, state1) = iter1
+        (elem2, state2) = iter2
+        if( elem1.degree <= elem2.degree) 
+            push!(_new.rootList, elem1)
+            iter1 = iterate(l1, state1)
+        else
+            push!(_new.rootList, elem2)
+            iter2 = iterate(l2, state2)
+        end
+    end
+    while (iter1 !== nothing)
+        (elem1, state1) = iter1 
+        push!(_new.rootList, elem1) 
+        iter1 = iterate(l1, state1) 
+    end
+    while (iter2 !== nothing)
+        (elem2, state2) = iter2
+        push!(_new.rootList, elem2)
+        iter2 = iterate(l2, state2)
+    end
+    h1.rootList = _new.rootList
+    adjust(h1)
+    return 
+end
+
+function push!(_heap::MutableBinomialHeap{T}, key) where T
+    handle = _heap.cumm_nodecount + 1
+    temp = MutableBinomialHeapNode(convert(T, key), handle)
+    insertATreeInHeap(_heap, temp)
+    _heap.nodecount = _heap.nodecount + 1
+    _heap.cumm_nodecount = _heap.cumm_nodecount + 1
+    return temp.handle
+end
+#################################################
+#
+#   interfaces
+#
+#################################################
+length(h::MutableBinomialHeap) = h.nodecount
+
+isempty(h::MutableBinomialHeap) = isempty(h.rootList)
+
+"""
+union!(h1::MutableBinomialHeap, h2::MutableBinomialHeap)
+
+merges heap `h2` into heap `h1`
+"""
+union!(h1::MutableBinomialHeap, h2::MutableBinomialHeap) = union!(h1, h2, true)
+
+
+function sizehint!(h::MutableBinomialHeap, s::Integer)
+    sizehint!(h.rootList, s)
+    return h
+end
+
+@inline top(heap::MutableBinomialHeap) =  top_with_handle(heap)[1]
+
+minimum(heap::MutableBinomialHeap) =  top_with_handle(heap)[1]
+"""
+top_with_handle(h::MutableBinomialHeap)
+
+Returns the minimum element of the heap `h` and its handle.
+"""
+function top_with_handle(h::MutableBinomialHeap) 
+    (val, node) = top_elem_with_handle(h)
+    return (val, node.handle)
+end
```

Comment:
This interface here could use a rewrite, I think. If the heap is empty, `top_elem_with_handle(h)` will return `nothing`, which causes this to fail with an obscure error. Also, since the compiler can't guarantee that `top_elem_with_handle` will return a 2-tuple, it needs all sorts of boilerplate code in this function which prevents all these 3 functions from inlining, despite the `@inline` notation.
Perhaps rewrite it such that `top_elem_with_handle` either raises an error, or returns a 2-tuple, and then make these three functions `@inline` and `@inbounds` for efficiency?

Reply:
yeah, I am on this

Reply:
By the way, `Base.generator` uses the following trick: Say `foo` is a function that can return either a 2-tuple or `nothing`. Then you can do:
```
val = foo()
if val === nothing
    # do stuff with nothing
else
    val = val::Tuple{Any,Any} # force compiler to know val is a 2-tuple
    return val[1] # compiler will now remove the boundscheck
end
```
This is just as efficient as rewriting `top_elem_with_handle` to never return `nothing`, but may be easier to write.